<launch>
    <arg name="x_pos" default="-2.0"/> 
    <arg name="y_pos" default="0.5"/>
    <arg name="z_pos" default="0.5"/>

    <!-- -2 0 for turtle -->
    <!-- -2 1 for house -->

    <!-- <arg name="world" default="turtle"/> -->
    <arg name="world" default="house"/>
    <!-- <arg name="world" default="gas_station"/> -->

    <arg name="mode" default="default"/>

    <!-- Argument to pass the Xacro file path -->
    <arg name="model" default="$(find rocker_bogie)/urdf/rocker_bogie.urdf.xacro" />

    <!-- Robot Description -->
    <param name="robot_description" 
           command="$(find xacro)/xacro --inorder $(arg model)" />

    <!-- Launch the robot_state_publisher to publish transforms -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen" />
    <!-- Launch the joint_state_publisher to publish joint states -->
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" output="screen" />
    <!-- Movement Publisher -->
    <node pkg="rocker_bogie" type="movement_publisher.py" name="movement_publisher" output="screen"/>
    <!-- Base Link Broadcaster -->
    <node pkg="rocker_bogie" type="base_link_broadcaster.py" name="base_link_broadcaster" output="screen"/>
    <!-- <node pkg="rocker_bogie" type="rocker_bogie_tf_publisher.py" name="rocker_bogie_tf_publisher" output="screen"/> -->

    <!-- Launch an empty Gazebo world -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="paused" value="false" />
        <arg name="use_sim_time" value="true" />  <!-- Sync with ROS clock -->
        <arg name="world_name" value="$(find rocker_bogie)/world/$(arg world).world" />
    </include>

    <!-- Spawn the rocker-bogie robot into Gazebo -->
    <node name="spawn_rocker_bogie" pkg="gazebo_ros" type="spawn_model"
          args="-urdf -param robot_description -model rocker_bogie -x $(arg x_pos) -y $(arg y_pos) -z $(arg z_pos)" output="screen" />

    <!-- Launch Control Node -->
    <node pkg="rocker_bogie" type="skid_steering_control.py" name="skid_steering_control" output="screen"/>
    
    <!-- SLAM: Gmapping -->
    <group if="$(eval arg('mode') == 'slam')">
         <include file="$(find rocker_bogie)/launch/slam.launch"/>
    </group>

    <!-- Navigation -->
    <group if="$(eval arg('mode') == 'nav')">
        <include file="$(find rocker_bogie)/launch/navigation.launch">
            <arg name="map_file" value="$(find rocker_bogie)/map/$(arg world)/map.yaml"/>
        </include>
    </group>

</launch>
